Add scala 3 contributing guide - MVP (#2194)
scala/docs.scala-lang: _overviews/scala3-contribution/procedures-testing.md

---
title: Testing Your Changes
type: section
description: This page describes test procedures in the Scala 3 compiler.
num: 10
previous-page: procedures-efficiency
next-page: procedures-checklist
---

It is important to add tests before a pull request, to verify that everythig is working as expected,
and act as proof of what is valid/invalid Scala code (in case it is broken in the future).

> In this section we will see the testign procedures in Scala 3.

Reviewer: ```suggestion
In this section we will see the testing procedures in Scala 3.
```

## Running all Tests

To run all tests of Scala 3, including for compiler, REPL, libraries and more, run the following in sbt:

```bash
$ sbt
sbt:scala3> scala3-bootstrapped/test
```

Often however it is not necessary to test everything if your changes are localised to one area,
we will see in the following sections the different kinds of tests, and how
to run individual tests.

## Compilation Tests

Compilation tests run the compiler over input files, using various settings. Input files
are found within the `tests/` directory at the root of the compiler repo.

> Test input files are categorised further by the virtue of placing them in the subdirectories

Reviewer: ```suggestion
Test input files are categorised further by placing them in the subdirectories
```

of the `tests/` directory. A small selection of test categories include:

- `tests/pos` – tests that should compile: pass if compiles successfully.
- `tests/neg` – should not compile: pass if fails compilation. Useful, e.g., to test an expected compiler error.
- `tests/run` – these tests not only compile but are also run.

### Naming and Running a Test Case

Tests are, by convention, named after the number of the issue they are fixing.
e.g. if you are fixing issue 101, then the test should be named `i101.scala`, for a single-file test,
or be within a directory called `i101/` for a multi-file test.

To run the test, invoke the sbt command `testCompilation i101` (this will match all tests with `"i101"` in
the name, so it is useful to use a unique name)

> The group tests – `pos`, `neg` etc – are defined in [CompilationTests]. If you want to run a group of tests, e.g.

Reviewer: ```suggestion
The test groups – `pos`, `neg`, etc. – are defined in [CompilationTests]. If you want to run a group of tests, e.g.
```

If you need more than one file to reproduce an issue, create a directory instead of a file
e.g. `tests/pos/i101/`, and put all the Scala files that are needed to reproduce the issue there.
There are two ways to organise the input files within:

**1: Requiring classpath dependency:** Sometimes issues require one file to be compiled after the other,
(e.g. if the issue only happens with a library dependency, like with Java interop). In this case,
the outputs of the first file compiled will be available to the next file compiled, available via the classpath.
This is called *separate compilation*.

To achieve this, within `tests/pos/i101/`, add a suffix `_n` to each file name, where `n` is an integer defining the
order in which the file will compile. E.g. if you have two files, `Lib.scala` and `Main.scala`, and you need them
compiled separately – Lib first, Main second, then name them `Lib_1.scala` and `Main_2.scala`.

**2: Without classpath dependency:** If your issue does not require a classpath dependency, your files can be compiled
in a single run, this is called *joint compilation*. In this case use file names without the `_n` suffix.

### Checking Program Output

> `tests/run` tests verify the runtime behaviour of a test case. The output is checked by invoking a main method

Reviewer: ```suggestion
`tests/run` tests verify the run-time behaviour of a test case. The output is checked by invoking a main method
```

on a class `Test`, this can be done with either
```scala
@main def Test: Unit = assert(1 > 0)
```
or
```scala
object Test extends scala.App:
  assert(1 > 0)
```

If your program also prints output, this can be compared against `*.check` files.
These contain the expected output of a program. Checkfiles are named after the issue they are checking,

> e.g. `tests/run/i101.check` will check both of `tests/run/i101.scala` and `tests/run/i101/`.

Reviewer: ```suggestion
e.g. `tests/run/i101.check` will check either `tests/run/i101.scala` or `tests/run/i101/`.
```

### Checking Compilation Errors

`tests/neg` tests verify that a file does not compile, and user-facing errors are produced. There are other neg
categories such as `neg-custom-args`, i.e. with `neg` prefixing the directory name. Test files in the `neg*`
categories require annotations for the lines where errors are expected. To do this add one `// error` token to the

> end of a line for each error expected. e.g. if there are three errors expected, the end of line should contain

Reviewer: ```suggestion
end of a line for each expected error. For example, if there are three expected errors, the end of the line should contain
```

`// error // error // error`.

You can verify the content of the error messages with a `*.check` file. These contain the expected output of the
compiler. Checkfiles are named after the issue they are checking,

> e.g. `i101.check` will check both of `tests/neg/i101.scala` and `tests/neg/i101/`.

Reviewer: ```suggestion
e.g. `i101.check` will check either `tests/neg/i101.scala` or `tests/neg/i101/`.
```

> *Note:* checkfiles are not required for the test to pass, however they do add more strong contraints that the errors

Reviewer: ```suggestion
*Note:* checkfiles are not required for the test to pass, however they do add stronger constraints that the errors
```
